### DISCUSSION TITLE
how do I turn the Result object table into mdast or hast?

### DISCUSSION THREAD

**Discussion by reypm** [remarkjs/remark#1218] [Q&A] [answered]
I was in the middle of write some custom code to parse a custom directive that is coming as markdown content and has the following syntax:
```
:::page-heading-list{min-level=2 max-level=3}
In this section // this should be the TOC header|title
:::
```
See the following example of markdown content:
```markdown
# this h1
## this h2
## this h2
### this h3
#### this h4
```
The idea behind this was to create a TOC for all the h2 (set by min-leve), h3 (set by max-level). On top of my head I should end up with something like:
```
 - [this h2](#this-h2)
 - [this h2](#this-h2)
 - [this h3](#this-h3)
```
Doing some research I found this two:
- [remark-toc](https://github.com/remarkjs/remark-toc)
- [mdast-util-toc](https://github.com/remarkjs/remark-toc)

This is my current code:
```tsx
const processPageHeadingList = (
  node: ContainerDirective | LeafDirective,
  index: number,
  parent: any
) => {
  const minLevel: number = node.attributes?.['min-level'] !== undefined ? parseInt(node.attributes?.['min-level']) : 0
  const maxLevel: number = node.attributes?.['max-level'] !== undefined ? parseInt(node.attributes?.['max-level']) : 0

  node.data = node.data ?? {}
  node.data.hName = 'nav'
  node.data.hProperties = {
    className: 'page-heading-list',
    ariaLabel: 'Page heading list'
  }

  let headings: any = []

  parent.children.map((element: any) => {
    if (element.type === 'heading' && element.depth <= maxLevel && element.depth >= minLevel) {
      headings.push(element)
    }
  })

  /** @type {import('mdast').Root} */
  const tocElement = {
    type: 'root',
    children: parent.children
  }

  const table = toc(tocElement)

  console.dir(table, { depth: 3 })
}

export default processPageHeadingList
```
and it partially works, however:

- is it possible to set a constraint to just start the TOC from the `min-level`?
- how do I turn the `Result` object `table` into `mdast` or `hast`?

Here is a runnable example for the above: https://codesandbox.io/s/remark-directive-forked-tjy5z6

**Comment by ChristianMurphy** [comment]
> is it possible to set a constraint to just start the TOC from the min-level?

I'm not sure what you mean by this.
There is not an option for this currently https://github.com/syntax-tree/mdast-util-toc#options
An option _could_ be added, but there would need to be some definition about why a min level is needed, and how it would work in practice.

> how do I turn the Result object table into mdast or hast?

Reading the documentation is a good starting point https://github.com/syntax-tree/mdast-util-toc#result 🙂 
`Result.map` is an MDAST `List` object.

MDAST can be turned into HAST with `mdast-util-to-hast` or its wrapper `remark-rehype`.

Or if you are adding additional headings through your remark directives or rehype plugins, you could also consider using `rehype-toc` https://github.com/JS-DevTools/rehype-toc

**Reply by reypm** [reply]
howdy @ChristianMurphy, 

> An option could be added, but there would need to be some definition about why a min level is needed, and how it would work in practice.

`min-level` in our universe is because, for some reason, I don't even know, we want to skip heading levels. For example exclude level1 (h1) and level4(and beyond). This is customizable and if you don't add them it will go from h1 to h6. See the following example of markdown content:

```
# this h1
## this h2
## this h2
### this h3
#### this h4
```

by using:

```
:::page-heading-list{min-level=2 max-level=3}
In this section // this should be the TOC header|title
:::
```

The resulting TOC should only contain: h2 and h3 ending up with something like this:

```
 - [this h2](#this-h2)
 - [this h2](#this-h2)
 - [this h3](#this-h3)
 ```

I believe I can use the existent `mathDepth`  for the `max-level` constraint but I am not sure how to add `min-level` as an `option` could you please provide me with an example?

> MDAST can be turned into HAST with mdast-util-to-hast or its wrapper remark-rehype

Everything coming out from these plugins like `processPageHeadingList` has to be turned into `hast`? I am not using `remark-rehype` if I add it should I forget about manually converting `mdast` into `hast`?

I partially get this working but should not `mdast-util-toc` generate URL's to these headings? Here is my "working" solution but missing links, can you check if I am missing something? https://codesandbox.io/s/remark-directive-forked-tjy5z6?file=/src/customPlugins/Shortcodes/UnorderedList.tsx:736-1284

One more thing, Typescript is driving me crazy with errors like:

>   TS2345: Argument of type 'Root' is not assignable to parameter of type 'Nodes'.

which is happening on this line: `const table = toc(tree, { depth: maxLevel })` because of `tree` being `Root` as per docs:

```
const tree: import('mdast').Root = {
    type: 'root',
    children: headings
  }
```

And another one is on this line `toHast(table.map)`:

> TS2345: Argument of type 'List | undefined' is not assignable to parameter of type 'MdastNode'. Type 'undefined' is not assignable to type 'MdastNode'.

I guess because `table.map` is type of List and not Node but then how I can convert it to `hast`? 🤔 

Again, thank you for your time and patience, I am still learning through the process.

**Reply by ChristianMurphy** [reply]
> min-level in our universe is because, for some reason, I don't even know, we want to skip heading levels. 

You may want to get some context on why before implementing it.
Otherwise you may implement an incorrect or unneeded feature.

> I believe I can use the existent mathDepth for the max-level constraint but I am not sure how to add min-level as an option could you please provide me with an example?

Here is where/how `maxDepth` is implemented. https://github.com/syntax-tree/mdast-util-toc/blob/3488516f7f91183fda72ebd443fe28bd24249f1c/lib/search.js#L128
How about you try implementing the feature, and ask questions along the way?

> Everything coming out from these plugins like processPageHeadingList has to be turned into hast? 

If you want HTML or React output, yes, at some point it needs to be turned into HAST.
In `react-markdown` that is implicitly/already a part of the pipeline, you probably don't need to add it again, _unless_ you need to leverage `node.data.hChildren`.

> I partially get this working but should not mdast-util-toc generate URL's to these headings? 

`mdast-util-toc` and `remark-toc` assume that the heading will have an `id` following the GitHub slug format.
These can be added to your HAST/HTML with rehype plugin https://github.com/rehypejs/rehype-slug.

> TS2345: Argument of type 'Root' is not assignable to parameter of type 'Nodes'.

This error doesn't show up in the sandbox you shared.
I'm assuming you are talking about this part:

```ts
const tree = {
  type: "root",
  children: headings
};

const table = toc(tree, { maxDepth: maxLevel, tight: true });
```

?
If so I'd recommend reading https://unifiedjs.com/learn/recipe/build-a-syntax-tree/ which gives an intro to generating strongly typed trees.
The issue here is that objects are widened by typescript by default. Instead of `type: 'root'` (which is valid), `type: string` will appear (which is not a valid MDAST node).
to avoid this, either:
* assign to an MDAST node type `const tree: Root = {`
* use a const assertion https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions
   ```ts
   const tree = {
     type: "root",
     children: headings
   } as const;
   ```

Your current sandbox also has a similar issue on this line
```ts
const table = toc(tree, { maxDepth: maxLevel, tight: true });
```

where

```
Type 'number' is not assignable to type '1 | 6 | 2 | 3 | 4 | 5'.ts(2322)
```

this is because you currently have `maxLevel` as a `number` type, which includes values which `toc` does not support (like `0`, `7`, `NaN`)

> ```
> TS2345: Argument of type 'List | undefined' is not assignable to parameter of type 'MdastNode'. Type 'undefined' is not assignable to type 'MdastNode'.
> ```

The `toc` plugin may not find any items to add to the table of contents.
If so it returns `undefined`, `undefined` is not a valid `MdastNode` so it correctly warns.
If a table of contents is not found, choose how you want to handle it, some common approaches are:
* skip adding the table of contents to the page
* throw an error

**Reply by reypm** [reply]
> You may want to get some context on why before implementing it.
> Otherwise you may implement an incorrect or unneeded feature.

You are correct, I am asking around why is this needed because it does not make any sense to me but I can be mistaken.
For now, my solution is to restrict what is being passed to `mdast-util-doc` aka building my own list of allowed
headings. See below:

```tsx
parent.children.map((element: any) => {
    if (
        element.type === "heading" &&
        element.depth <= maxLevel && // here
        element.depth >= minLevel // here
    ) {
        const listItemValue = element.children[0].hasOwnProperty("children")
            ? element.children[0].children[0].value
            : element.children[0].value;

        const link: Link = {
            type: 'link',
            url: "#" + listItemValue.replace(/\s+/g, "-").toLowerCase(),
            title: listItemValue,
            children: [{type: 'text', value: listItemValue}]
        };

        headings.push(link);
    }
});
```

> How about you try implementing the feature, and ask questions along the way?

Sure I can contribute to this repo if that is what you are asking me but I need to find the why first so I can fight
back with you and the community heh.

> If you want HTML or React output, yes, at some point it needs to be turned into HAST.

Got it.

> In react-markdown that is implicitly/already a part of the pipeline, you probably don't need to add it again, unless
> you need to leverage node.data.hChildren.

The above confuses me, let's say I am not messing up with `node.data.hChildren` at all, what my code should do? Just
replace the content for what I want and let it go?

> mdast-util-toc and remark-toc assume that the heading will have an id following the GitHub slug format.
> These can be added to your HAST/HTML with rehype plugin rehypejs/rehype-slug.

Since the headings on the markdown do not have any link this is how I am building it (see the above snippet for better
understanding):

```tsx
const link: Link = {
    type: 'link',
    url: "#" + listItemValue.replace(/\s+/g, "-").toLowerCase(),
    title: listItemValue,
    children: [{type: 'text', value: listItemValue}]
};
```
that should yield something like: `[alpha](https://example.com "bravo")` if I am not mistaken, is that what you mean?

And as per the Typescript is already getting out my gray hair Heh 😄

**Reply by ChristianMurphy** [reply]
> Sure I can contribute to this repo if that is what you are asking me but I need to find the why first

Do find out why first, then consider opening a separate issue/discussion on the feature.
Usually it takes some discussion with the maintainer team to figure out if the new option is a good fit and what it should look like.

> The above confuses me, let's say I am not messing up with node.data.hChildren at all, what my code should do? Just replace the content for what I want and let it go?

Yes? If you are working with an MDAST tree (which it looks like you are) you can directly add/splice/replace the `List` into the content where you want it to go, and `react-markdown` will take care of the rest.

> Since the headings on the markdown do not have any link this is how I am building it 

I'm not sure I follow.
The anchor part of the URL (anything starting with `#`) needs to refer to an HTML element with an `id` attribute that matches the anchor in the URL.
You don't have `id` attributes added to the heading currently.
Until you do, no anchor links will work, no matter how you regex replace the URL.

https://github.com/rehypejs/rehype-slug can add `id` attributes to headings.

**Reply by reypm** [reply]
> I'm not sure I follow.
> The anchor part of the URL (anything starting with #) needs to refer to an HTML element with an id attribute that
> matches the anchor in the URL.
> You don't have id attributes added to the heading currently.
> Until you do, no anchor links will work, no matter how you regex replace the URL.

What I mean is that I can write the following markdown:

```
::page-heading-list[In this section]{min-level=1 max-level=3}

- [Item 0](#item-0) // if I click on this one it will go to the h1 #item 0
- [Item 1](#item-1) // if I click on this one it will go to the h2 #item 1

# item 0

## item 1
```

> rehypejs/rehype-slug can add id attributes to headings.

I am using it and it is adding the `id` to each heading element.

<img width="515" alt="image" src="https://github.com/remarkjs/remark/assets/627886/8866eb67-c8c3-4fed-a9eb-db8219302045">

but seems that I am missing something else, probably I am building the wrong elements here or I haven't
understood how the plugin works and what it needs as input.

**Reply by ChristianMurphy** [reply]
What is your question?
If you have ids on the headings, the table of contents has anchor references to the headings, and the page jumps to the correct location when you click on the link.
Then everything is working?

**Reply by reypm** [reply]
Can you check really quick here https://codesandbox.io/s/remark-directive-forked-tjy5z6?file=/src/customPlugins/Shortcodes/UnorderedList.tsx ?? I am able to print out the toc but no links and I am not sure why because the headings has ids

**Reply by reypm** [reply]
> Do find out why first, then consider opening a separate issue/discussion on the feature.

Got my answer: they don't want to include the page title which is an h1. However, I think this is achievable by using `options.skip` I will check on this

**Reply by ChristianMurphy** [reply, marked as answer]
> Can you check really quick here https://codesandbox.io/s/remark-directive-forked-tjy5z6?file=/src/customPlugins/Shortcodes/UnorderedList.tsx ?? I am able to print out the toc but no links and I am not sure why because the headings has ids

The answer remains the same as before

> If you are working with an MDAST tree (which it looks like you are) you can directly add/splice/replace the List into the content where you want it to go, and react-markdown will take care of the rest.

Don't run `toHast` on `table.map`, directly add `table.map` into the MDAST document.

**Reply by reypm** [reply]
I have chosen the latest as the answer but all the information contained here was really helpful to me and helped me to understand a little bit more how things work and what I was doing wrong.
